lib/irq.h

```diff
@@ -19,61 +19,85 @@ extern "C" {
 /** \defgroup irq Interrupt Handling Interfaces
  *  @{ */
 
+#include <metal/list.h>
 #include <stdlib.h>
 
 /** IRQ handled status */
 #define METAL_IRQ_NOT_HANDLED 0
 #define METAL_IRQ_HANDLED     1
 
+/** IRQ ANY ID */
+#define METAL_IRQ_ANY         (-1)
+
+/** IRQ state macro which will be passed to metal irq set state function
+ *  to indicate which state the caller want the IRQ to change to.
+ */
+#define METAL_IRQ_DISABLE    0U
+#define METAL_IRQ_ENABLE     1U
+
+struct metal_irq_controller;
+
 /**
  * @brief	type of interrupt handler
- * @param[in]   irq interrupt id
- * @param[in]	priv private data
- * @return      irq handled status
+ * @param[in]	irq interrupt id
+ * @param[in]	arg argument to pass to the handler
+ * @return	irq handled status
  */
-typedef int (*metal_irq_handler) (int irq, void *priv);
+typedef int (*metal_irq_handler) (int irq, void *arg);
 
-struct metal_device;
+/**
+ * @brief	type of interrupt controller to set irq enable
+ * @param[in]   irq_cntr pointer to interrupt controller
+ * @param[in]   irq interrupt id
+ * @param[in]   enable IRQ state
+ */
+typedef void (*metal_irq_set_enable) (struct metal_irq_controller *irq_cntr,
+				     int irq, unsigned int enable);
+
+/** Libmetal interrupt structure */
+struct metal_irq {
+	metal_irq_handler hd; /**< Interrupt handler */
+	void *arg; /**< Argument to pass to the interrupt handler */
+};
+
+/** Libmetal interrupt controller structure */
+struct metal_irq_controller {
```

Comment:
It's better to split metal_irq_controller and related API to another header file(e.g. irqchip.h or irqcontroller.h).
1.metal_irq_register/metal_irq_enable is for application
2.metal_irq_register_controller is for porting layer